File: sdk/typescript/src/a2a/invitations.ts
/** Organization-managed A2A invitations. */

import type { HttpTransport } from "../_http.js";

export type A2AInvitationStatus =
  | "pending"
  | "awaiting_verification"
  | "accepted"
  | "declined"
  | "revoked"
  | "expired";

export type A2AInvitationEmailStatus =
  | "not_requested"
  | "pending"
  | "sent"
  | "failed"
  | "indeterminate";

export interface A2AInvitation {
  id: string;
  issuerOrganizationId: string;
  inviterEmail: string | null;
  peerAgentHandles: string[];
  recipientEmail: string | null;
  status: A2AInvitationStatus;
  emailStatus: A2AInvitationEmailStatus;
  emailSentAt: string | null;
  inviteeIdentityId: string | null;
  inviteeAgentHandle: string | null;
  inviteeOrganizationId: string | null;
  expiresAt: string;
  acceptedAt: string | null;
  declinedAt: string | null;
  revokedAt: string | null;
  createdAt: string;
  updatedAt: string;
}

export interface A2AInvitationCreateOptions {
  peerAgentHandles: string[];
  recipientEmail?: string;
  expiresInSeconds?: number;
}

export interface A2AInvitationCreateResult extends A2AInvitation {
  /** Present only when the invitation is not bound to a recipient email. */
  invitationToken?: string;
  /** Shareable acceptance link, when returned by the server. */
  invitationUrl?: string;
  /** Present only when the invitation is not bound to a recipient email. */
  agentHandoffPrompt?: string;
}

export interface A2AInvitationListOptions {
  status?: A2AInvitationStatus;
  cursor?: string;
  limit?: number;
}

export interface A2AInvitationPage {
  items: A2AInvitation[];
  nextCursor: string | null;
}

export interface A2AInvitationAcceptResult {
  invitationId: string;
  status: "accepted";
  inviteeIdentityId: string;
  inviteeAgentHandle: string;
  peerAgentHandles: string[];
  acceptedAt: string;
}

type Raw = Record<string, any>;

/** Raised when an A2A invitation link is invalid for the configured site. */
export class A2AInvitationParseError extends Error {
  constructor(message: string) {
    super(message);
    this.name = "A2AInvitationParseError";
  }
}

const malformedEscape = /%(?![0-9A-Fa-f]{2})/;
const invitationTokenPattern = /^a2ai_[A-Za-z0-9_-]{43}$/;
const maxInvitationInputBytes = 2048;
const officialInvitationOriginsByApiOrigin = new Map([
  ["https://api.inkbox.ai", "https://inkbox.ai"],
  ["https://api.beta.inkbox.ai", "https://beta.inkbox.ai"],
  ["https://api.development.inkbox.ai", "https://development.inkbox.ai"],
]);

function invitationOriginAllowed(invitationUrl: URL, configuredUrl: URL): boolean {
  return invitationUrl.origin === configuredUrl.origin
    || officialInvitationOriginsByApiOrigin.get(configuredUrl.origin)
      === invitationUrl.origin;
}

/**
 * Return the raw token from an A2A invitation link or raw token.
 * Links require HTTPS, except configured localhost/127.0.0.1 development URLs.
 */
export function extractA2AInvitationToken(
  value: string,
  baseUrl = "https://inkbox.ai",
): string {
  if (!value || new TextEncoder().encode(value).byteLength > maxInvitationInputBytes) {
    throw new A2AInvitationParseError("The A2A invitation link or token is invalid.");
  }
  const candidate = value.trim();
  if (invitationTokenPattern.test(candidate)) return candidate;
  if (
    !candidate.includes("://")
    && !candidate.toLowerCase().startsWith("http:")
    && !candidate.toLowerCase().startsWith("https:")
  ) {
    throw new A2AInvitationParseError("The A2A invitation link or token is invalid.");
  }
  if (!/^https?:\/\//i.test(candidate)) {
    throw new A2AInvitationParseError("The A2A invitation link is invalid.");
  }

  let invitationUrl: URL;
  let configuredUrl: URL;
  try {
    invitationUrl = new URL(candidate);
    configuredUrl = new URL(baseUrl);
  } catch {
    throw new A2AInvitationParseError("The A2A invitation link is invalid.");
  }
  if (
    !["http:", "https:"].includes(invitationUrl.protocol)
    || !["http:", "https:"].includes(configuredUrl.protocol)
    || invitationUrl.username
    || invitationUrl.password
  ) {
    throw new A2AInvitationParseError("The A2A invitation link is invalid.");
  }
  if (
    configuredUrl.protocol === "http:"
    && configuredUrl.hostname !== "localhost"
    && configuredUrl.hostname !== "127.0.0.1"
  ) {
    throw new A2AInvitationParseError("The configured site URL is invalid.");
  }
  if (!invitationOriginAllowed(invitationUrl, configuredUrl)) {
    throw new A2AInvitationParseError(
      "The A2A invitation link does not match the configured site.",
    );
  }
  const fragmentStart = candidate.indexOf("#");
  const beforeFragment = candidate.slice(
    0,
    fragmentStart < 0 ? undefined : fragmentStart,
  );
  if (
    invitationUrl.pathname !== "/console/a2a/invitations/accept"
    || beforeFragment.includes("?")
  ) {
    throw new A2AInvitationParseError("The A2A invitation link is invalid.");
  }
  const fragment = invitationUrl.hash.slice(1);
  if (malformedEscape.test(fragment)) {
    throw new A2AInvitationParseError("The A2A invitation link is invalid.");
  }
  const fields = [...new URLSearchParams(fragment).entries()];
  if (
    fields.length !== 1
    || fields[0]?.[0] !== "token"
    || !invitationTokenPattern.test(fields[0]?.[1] ?? "")
  ) {
    throw new A2AInvitationParseError("The A2A invitation link is invalid.");
  }
  return fields[0][1];
}

function parseInvitation(raw: Raw): A2AInvitation {
  return {
    id: raw.id,
    issuerOrganizationId: raw.issuer_organization_id,
    inviterEmail: raw.inviter_email ?? null,
    peerAgentHandles: raw.peer_agent_handles,
    recipientEmail: raw.recipient_email ?? null,
    status: raw.status,
    emailStatus: raw.email_status,
    emailSentAt: raw.email_sent_at ?? null,
    inviteeIdentityId: raw.invitee_identity_id ?? null,
    inviteeAgentHandle: raw.invitee_agent_handle ?? null,
    inviteeOrganizationId: raw.invitee_organization_id ?? null,
    expiresAt: raw.expires_at,
    acceptedAt: raw.accepted_at ?? null,
    declinedAt: raw.declined_at ?? null,
    revokedAt: raw.revoked_at ?? null,
    createdAt: raw.created_at,
    updatedAt: raw.updated_at,
  };
}

export class A2AInvitationsResource {
  constructor(
    private readonly http: HttpTransport,
    private readonly baseUrl = "https://inkbox.ai",
  ) {}

  async create(options: A2AInvitationCreateOptions): Promise<A2AInvitationCreateResult> {
    const raw = await this.http.post<Raw>("/a2a/invitations", {
      peer_agent_handles: options.peerAgentHandles,
      ...(options.recipientEmail !== undefined
        ? { recipient_email: options.recipientEmail }
        : {}),
      ...(options.expiresInSeconds !== undefined
        ? { expires_in_seconds: options.expiresInSeconds }
        : {}),
    });
    return {
      ...parseInvitation(raw),
      ...(raw.invitation_token !== undefined
        ? { invitationToken: raw.invitation_token }
        : {}),
      ...(typeof raw.invitation_url === "string"
        ? { invitationUrl: raw.invitation_url }
        : {}),
      ...(raw.agent_handoff_prompt !== undefined
        ? { agentHandoffPrompt: raw.agent_handoff_prompt }
        : {}),
    };
  }

  async list(options: A2AInvitationListOptions = {}): Promise<A2AInvitationPage> {
    const raw = await this.http.get<Raw>("/a2a/invitations", {
      status: options.status,
      cursor: options.cursor,
      limit: options.limit ?? 50,
    });
    return {
      items: (raw.items ?? []).map(parseInvitation),
      nextCursor: raw.next_cursor ?? null,
    };
  }

  async get(invitationId: string): Promise<A2AInvitation> {
    return parseInvitation(
      await this.http.get<Raw>(`/a2a/invitations/${encodeURIComponent(invitationId)}`),
    );
  }

  async revoke(invitationId: string): Promise<A2AInvitation> {
    return parseInvitation(
      await this.http.post<Raw>(
        `/a2a/invitations/${encodeURIComponent(invitationId)}/revoke`,
      ),
    );
  }

  async accept(invitation: string): Promise<A2AInvitationAcceptResult> {
    const invitationToken = extractA2AInvitationToken(invitation, this.baseUrl);
    const raw = await this.http.post<Raw>("/a2a/invitations/accept", {
      invitation_token: invitationToken,
    });
    return {
      invitationId: raw.invitation_id,
      status: raw.status,
      inviteeIdentityId: raw.invitee_identity_id,
      inviteeAgentHandle: raw.invitee_agent_handle,
      peerAgentHandles: raw.peer_agent_handles,
      acceptedAt: raw.accepted_at,
    };
  }
}

File: sdk/rust/src/a2a/invitations.rs
//! Parsing helpers for shareable A2A invitation links.

use url::{form_urlencoded, Url};

const DEFAULT_BASE_URL: &str = "https://inkbox.ai";
const MAX_INVITATION_INPUT_BYTES: usize = 2048;

/// An A2A invitation link is invalid for the configured site.
#[derive(Debug, Clone, Copy, PartialEq, Eq, thiserror::Error)]
#[error("the A2A invitation link is invalid")]
pub struct A2AInvitationParseError;

fn malformed_percent_encoding(value: &str) -> bool {
    let bytes = value.as_bytes();
    let mut index = 0;
    while index < bytes.len() {
        if bytes[index] == b'%' {
            if index + 2 >= bytes.len()
                || !bytes[index + 1].is_ascii_hexdigit()
                || !bytes[index + 2].is_ascii_hexdigit()
            {
                return true;
            }
            index += 3;
        } else {
            index += 1;
        }
    }
    false
}

/// Return the raw token from an A2A invitation link or raw token.
///
/// Links must use HTTPS (HTTP only for localhost/127.0.0.1), the configured
/// site's exact origin, and the public invitation acceptance path. Errors
/// never include the capability value.
pub fn extract_a2a_invitation_token(value: &str) -> Result<String, A2AInvitationParseError> {
    extract_a2a_invitation_token_with_base_url(value, DEFAULT_BASE_URL)
}

/// Return the raw token using an explicit site base URL.
pub fn extract_a2a_invitation_token_with_base_url(
    value: &str,
    base_url: &str,
) -> Result<String, A2AInvitationParseError> {
    if value.is_empty() || value.len() > MAX_INVITATION_INPUT_BYTES {
        return Err(A2AInvitationParseError);
    }
    let candidate = value.trim();
    if valid_token(candidate) {
        return Ok(candidate.to_owned());
    }
    if !candidate.contains("://")
        && !candidate.to_ascii_lowercase().starts_with("http:")
        && !candidate.to_ascii_lowercase().starts_with("https:")
    {
        return Err(A2AInvitationParseError);
    }
    if !candidate.to_ascii_lowercase().starts_with("http://")
        && !candidate.to_ascii_lowercase().starts_with("https://")
    {
        return Err(A2AInvitationParseError);
    }

    let invitation = Url::parse(candidate).map_err(|_| A2AInvitationParseError)?;
    let configured = Url::parse(base_url).map_err(|_| A2AInvitationParseError)?;
    let configured_http_allowed = configured.scheme() != "http"
        || matches!(configured.host_str(), Some("localhost" | "127.0.0.1"));
    if !matches!(invitation.scheme(), "http" | "https")
        || !matches!(configured.scheme(), "http" | "https")
        || !configured_http_allowed
        || !invitation.username().is_empty()
        || invitation.password().is_some()
        || invitation.origin() != configured.origin()
        || invitation.path() != "/console/a2a/invitations/accept"
        || invitation.query().is_some()
    {
        return Err(A2AInvitationParseError);
    }
    let fragment = invitation.fragment().ok_or(A2AInvitationParseError)?;
    if malformed_percent_encoding(fragment) {
        return Err(A2AInvitationParseError);
    }
    let fields: Vec<_> = form_urlencoded::parse(fragment.as_bytes()).collect();
    if fields.len() != 1 || fields[0].0 != "token" || !valid_token(&fields[0].1) {
        return Err(A2AInvitationParseError);
    }
    Ok(fields[0].1.to_string())
}

fn valid_token(value: &str) -> bool {
    value.len() == 48
        && value.starts_with("a2ai_")
        && value[5..]
            .bytes()
            .all(|byte| byte.is_ascii_alphanumeric() || matches!(byte, b'_' | b'-'))
}

#[cfg(test)]
mod tests {
    use super::*;
    use serde::Deserialize;

    #[derive(Deserialize)]
    struct Vector {
        base_url: String,
        input: String,
        token: Option<String>,
    }

    #[derive(Deserialize)]
    struct Vectors {
        valid: Vec<Vector>,
        invalid: Vec<Vector>,
    }

    #[test]
    fn accepts_raw_tokens_and_exact_origin_links() {
        assert_eq!(
            extract_a2a_invitation_token("a2ai_AAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAA")
                .unwrap(),
            "a2ai_AAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAA"
        );
        assert_eq!(
            extract_a2a_invitation_token_with_base_url(
                "https://inkbox.ai/console/a2a/invitations/accept#token=a2ai_AAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAA",
                "https://inkbox.ai/api/v1",
            )
            .unwrap(),
            "a2ai_AAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAA"
        );
    }

    #[test]
    fn rejects_untrusted_or_ambiguous_links_without_echoing_them() {
        for value in [
            "https://evil.example/console/a2a/invitations/accept#token=a2ai_AAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAA",
            "https://inkbox.ai/console/a2a/invitations/accept?token=a2ai_AAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAA",
            "https://inkbox.ai/console/a2a/invitations/accept#token=a&token=b",
            "https://inkbox.ai/console/a2a/invitations/accept#token=%ZZ",
            "https://inkbox.ai/a2a/invitations/accept#token=a2ai_AAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAA",
        ] {
            let error = extract_a2a_invitation_token_with_base_url(value, "https://inkbox.ai")
                .unwrap_err();
            assert!(!error
                .to_string()
                .contains("a2ai_AAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAA"));
        }
    }

    #[test]
    fn shared_invitation_input_vectors_match() {
        let vectors: Vectors = serde_json::from_str(include_str!(
            "../../../../tests/fixtures/a2a_invitation_inputs.json"
        ))
        .unwrap();
        for item in vectors.valid {
            assert_eq!(
                extract_a2a_invitation_token_with_base_url(&item.input, &item.base_url).unwrap(),
                item.token.unwrap()
            );
        }
        for item in vectors.invalid {
            let error = extract_a2a_invitation_token_with_base_url(&item.input, &item.base_url)
                .unwrap_err();
            assert!(!error
                .to_string()
                .contains("a2ai_AAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAA"));
        }
    }

    #[test]
    fn rejects_oversized_input() {
        assert!(extract_a2a_invitation_token(&"x".repeat(2049)).is_err());
    }
}
